### ISSUE TITLE
What's the reasonable training curve for different training stage, like pertaining on EgoDex dataset and HE dataset, mid train on HE dataset, and posttrain on the real and simple dataset separately?

### ISSUE THREAD

**Issue by Jeff09** [physical-superintelligence-lab/Psi0#41]

**Comment by songlin** [comment]
Hi @Jeff09,

For pre-training on EgoDEX, the training loss, action L1 loss, and action accuracy converge around 0.5, 0.004, and 85% after 200K steps with a global step size of 1024 and a constant learning rate of 1e-4. And the following pre-training on the HE dataset converges at similar value ranges, except that the L1 action loss is smaller. Which could attribute to its smaller data size.

For mid-training on the HE dataset in the joint space, after 30K steps of training with a global batch size of 2048 and a constant learning rate of 1e-4 (with VLM backbone frozen).  The eval loss converges around 3. Both L1 losses of the left and right hands converge around 0.11 (rad). And L1 losses of the left arm and right arm converge around 0.004 and 0.003, respectively.

For Fine-tuning, the loss curve depends very much on the task. In general, for all 8 real tasks released by Psi0, the eval losses should converge at 40000 steps with a global batch size of 128.  For example,

<img width="1972" height="690" alt="Image" src="https://github.com/user-attachments/assets/be181892-5fc6-4fc5-9fce-39965946fc95" />

Best, 
Songlin

**Comment by Jeff09** [comment]
Hi @songlin,

Thank you for the detailed information.

I’m currently trying to reproduce the reported results by loading the provided checkpoint on the EgoDex dataset:

> For pre-training on EgoDEX, the training loss, action L1 loss, and action accuracy converge around 0.5, 0.004, and 85%.

However, my results diverge significantly from these values. Here is what I obtained:
`{
  'loss': 11.2989,
  'action_accuracy': 0.1026,
  'action_ce_loss': 11.8783,
  'non_action_ce_loss': 0.0,
  'supervised_action_tokens': 39,
  'supervised_non_action_tokens': 2,
  'pred_action_token_rate': 1.0,
  'pixel_values_shape': (1, 480, 1536),
  'image_grid_thw': [[1, 16, 30]]
}`

I’ve also attached a test script that I used to generate these metrics for reference:

[eval_psi_exact_onebatch.py](https://github.com/user-attachments/files/27268691/eval_psi_exact_onebatch.py)

Could you please help identify what might be going wrong here?

**Comment by songlin** [comment]
Hi @Jeff09, that's unusual. I'll check back with you when I am available to debug.

**Comment by songlin** [comment]
Hi @Jeff09,

Apologies for the delayed reply. After carefully reviewing your code, I believe that neither token accuracy nor cross-entropy loss is appropriate for evaluating the model at test time. Both metrics are tied to teacher-forcing during training, and as a result, they don’t reflect true performance during autoregressive inference. In particular, token accuracy at test time is typically much lower due to exposure bias [1].

Instead, I recommend evaluating the pre-trained model using the decoded L1 loss, which better captures the quality of the model’s actual predictions. I’ve included a [sample script](https://github.com/physical-superintelligence-lab/Psi0/blob/main/scripts/test/test_pretrain_fast_egodex.py) for reference.

Using this approach, I evaluated the pre-trained model over 50 iterations and obtained an average L1 error of 0.00923. In my view, this indicates reasonably good predictive performance on EgoDex, given its data quality.

[1] He, Tianxing, et al. “Exposure Bias versus Self-Recovery: Are Distortions Really Incremental for Autoregressive Text Generation?” Proceedings of EMNLP 2021.

Best,
Songlin

**Comment by Jeff09** [comment]
Hi @songlin, 

Thank you for your clarification! I have got the similar result as yours using the provided script.

**Comment by Jeff09** [comment]
I evaluated several pre-training checkpoints on the EgoDex dataset using the decoded L1 loss metric. Here are some interesting findings.</p><p>The mean L1 error over 50 samples is consistently around <code inline="">0.01</code> across all checkpoints — including the checkpoint trained only on the HE dataset.</p>
PRETRAIN_CHECKPOINT | Mean L1 Error (50 samples)
-- | --
psi0/pre.abl.ego.10per.2602021632.46k | 0.0106
psi0/pre.abl.only.he.2512311516.48k | 0.0117
psi0/pre.fast.egodex.2512241941.ckpt200k | 0.0110
psi0/pre.fast.1by1.2601091803.ckpt.ego200k.he30k | 0.0106

<p>

This suggests that the decoded L1 loss on EgoDex may not be sufficiently sensitive to distinguish checkpoint quality differences in this setting.</p>

**Comment by songlin** [comment]
Thanks for sharing these insightful findings. We agree that this is an important consideration, which is why we report task completion rate in our ablation studies, as it better aligns with the objectives of pre-training. That said, we are actively exploring alternative metrics that could serve as more reliable proxy indicators for studying scaling laws in pre-training.
